packages/flutter/lib/src/widgets/async.dart

```diff
@@ -0,0 +1,256 @@
+      builder(context, currentSummary);
+}
+
+/// Base class for Widgets that build themselves based on interaction with
+/// a specified [stream].
+///
+/// The widget's state maintains a summary of the interaction so far. The
+/// summary is defined by sub-classes.
+///
+/// [T] type of stream events.
+/// [S] type of summary.
+abstract class StreamFold<T, S> extends StatefulWidget {
+  final Stream<T> stream;
+
+  StreamFold({Key key, this.stream}) : super(key: key);
+
+  /// Returns the initial summary.
+  S initial();
+
+  /// Returns an updated version of the [current] summary reflecting that we
+  /// are now connected to a stream.
+  ///
+  /// The default implementation returns [current] as is.
+  S onConnecting(S current) => current;
+
+  /// Returns an updated version of the [current] summary following an event.
+  S onData(S current, T data);
+
+  /// Returns an updated version of the [current] summary following an error.
+  ///
+  /// The default implementation returns [current] as is.
+  S onError(S current, dynamic error) => current;
+
+  /// Returns an updated version of the [current] summary following stream
+  /// termination.
+  ///
+  /// The default implementation returns [current] as is.
+  S onDone(S current) => current;
+
+  /// Returns an updated version of the [current] summary reflecting that we
+  /// are no longer connected to a stream.
+  ///
+  /// The default implementation returns [current] as is.
+  S onDisconnecting(S current) => current;
+
+  /// Returns a Widget based on the [currentSummary].
+  Widget build(BuildContext context, S currentSummary);
+
+  @override
+  State<StreamFold<T, S>> createState() => new _StreamFoldState<T, S>();
+}
+
+class _StreamFoldState<T, S> extends State<StreamFold<T, S>> {
+  StreamSubscription<T> _subscription;
+  S _summary;
+
+  @override
+  void initState() {
+    super.initState();
+    _summary = config.initial();
+    _subscribe();
+  }
+
+  @override
+  void didUpdateConfig(StreamFold<T, S> oldConfig) {
+    if (!identical(oldConfig.stream, config.stream)) {
+      _unsubscribe();
+      _subscribe();
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) => config.build(context, _summary);
+
+  @override
+  void dispose() {
+    _unsubscribe();
+    super.dispose();
+  }
+
+  void _subscribe() {
+    if (config.stream != null) {
+      _subscription = config.stream.listen((T event) {
+        setState(() {
+          _summary = config.onData(_summary, event);
+        });
+      }, onError: (dynamic e) {
+        setState(() {
+          _summary = config.onError(_summary, e);
+        });
+      }, onDone: () {
+        setState(() {
+          _summary = config.onDone(_summary);
+        });
+      });
+      _summary = config.onConnecting(_summary);
+    }
+  }
+
+  void _unsubscribe() {
+    if (_subscription != null) {
+      _subscription.cancel();
+      _subscription = null;
+      _summary = config.onDisconnecting(_summary);
```

Comment:
is there any particular reason to call this during dispose()? Maybe limit this to didUpdateConfig?